react_people-table-advanced Develop

DemetriodosAnjos opened the pull request:
[DEMO LINK] - https://DemetriodosAnjos.github.io/react_people-table-advanced/

DemetriodosAnjos committed:
/ MARCA: fim-etapa-ajustes-cores

DemetriodosAnjos committed:
/ MARCA: ajuste-layout-responsivo - tabela is-8-desktop e filtros is-4-desktop

DemetriodosAnjos committed:
Alterado a cor do Botão All do painel Filters

DemetriodosAnjos committed:
feat(people): selecionar linha ao clicar e manter compatibilidade de rotas

Ao clicar no nome, atualiza o hash para #/people/:slug e marca a linha como selecionada sem navegar para a página de detalhes. currentSelectedSlugFromHash prioriza rota /people/:slug com fallback para ?selected=. Mantém cores por sexo e compatibilidade com ordenação e query params.

mateacademy-ai-mentor commented on src/components/Navbar.tsx:
```diff
@@ -1,24 +1,40 @@
-export const Navbar = () => {
+import React from 'react';
+import { NavLink } from 'react-router-dom';
```
This file uses `NavLink` but doesn't access the current location. The task requires preserving existing search params when navigating within the People page (e.g., clicking the People link). Import `useLocation` from 'react-router-dom' so you can read `location.search` and append it to the People link. Example approach: `const location = useLocation();` then use `to={
  `/people${location.search}`
}` for the People NavLink. Without this, clicking the People link will drop any existing query/search params — this violates the requirement to keep search params when navigating within the People page.

mateacademy-ai-mentor commented on src/components/Navbar.tsx:
```diff
@@ -1,24 +1,40 @@
-            aria-current="page"
-            className="navbar-item has-background-grey-lighter"
-            href="#/people"
+          <NavLink
+            to="/people"
+            end
+            className={({ isActive }) =>
+              `navbar-item${isActive ? ' has-background-grey-lighter' : ''}`
+            }
+            data-cy="nav-people"
           >
             People
```
The People NavLink (to="/people") does not preserve current URL search params. Per the requirements, clicking the People link should keep existing search params. Modify this NavLink to include the current `location.search` (from `useLocation`) — e.g. `to={
  `/people${location.search}`
}`. This ensures that the search params currently in the URL are preserved when the user clicks the People link.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
+
+  useEffect(() => {
+    const onHashChange = () => {
+      const params = new URLSearchParams(
```
Right now you build URLSearchParams using window.location.hash.split('?')[1] here. Other components in the repo use a small helper that extracts the query part (and includes the leading `?` slice). Consider centralizing/parsing hash search params through a shared helper (e.g., `readSearchParamsFromHash`) so parsing is consistent across `PeoplePage`, `PeopleTable`, and `PeopleFilters`. This reduces duplication and the risk of subtle parsing differences when you change other code later.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
+    return () => window.removeEventListener('hashchange', onHashChange);
+  }, []);
+
+  const filteredPeople = useMemo(() => {
```
`PeoplePage` computes `filteredPeople` here using the query/sex/century filters. However `PeopleTable` also re-applies the same filtering and additionally performs sorting. This duplicates filter logic into two places which can cause maintenance problems. Decide on one single source of truth: either
1. Keep filtering & sorting in `PeopleTable` and pass the unfiltered `people` list from `PeoplePage`, or
2. Keep `PeoplePage` filtering and remove filtering from `PeopleTable`, making `PeopleTable` only responsible for sorting and rendering.
Either approach is valid, but choose one to avoid duplicated logic and potential inconsistencies (the filtering logic lives at lines where `.filter(p => ...)` is applied).

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
+              !loading &&
+              people &&
+              filteredPeople.length > 0 && (
+                <PeopleTable
```
You render `<PeopleTable>` only when `filteredPeople.length > 0`. That hides the table when the current filter produces zero results (you display a `no-results` message instead). This is acceptable, but confirm this is the intended UX — if you prefer the table to render with zero rows (so headers remain visible) then render `PeopleTable` even when result count is 0 and let it show an empty body. If you keep the current behavior, document the decision so future maintainers know why the table is omitted when there are no rows.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,275 @@
+    history.replaceState(null, '', newHash);
+    window.dispatchEvent(new HashChangeEvent('hashchange'));
+  }
+}
```
applySortExplicit uses an extra `sortCycle` param and always sets `order` (including for ascending). The task requires a 3-state cycle (ascending -> descending -> none) using `sort` and `order=desc` only; when ascending there should be no `order` param and there should be no `sortCycle` param in the URL. Please replace the current logic with the simpler sequence:
1. If current `sort` !== field: set `sort=<field>` and remove `order` (ascending state represented by `sort` only).
2. Else if `sort` === field and there is no `order`: set `order=desc` (descending).
3. Else if `sort` === field and `order === 'desc'`: remove both `sort` and `order` (clear sorting).
After changing params, update the hash exactly as you already do and dispatch the hashchange event. This will satisfy the URL param requirements in the description.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,275 @@
+      params.set('sortCycle', String(Math.min(currentCycle + 1, 2)));
+    } else {
+      params.set('order', orderClicked);
+      params.set('sortCycle', String(Math.min(currentCycle + 1, 2)));
```
This block sets `order` to `orderClicked` and writes `sortCycle` into the URL (both here and below). As explained above, the requirement specifically asks to add `order=desc` only when descending and to not include `sort` and `order` when sorting is disabled. Remove any `sortCycle` usage and only set `order='desc'` for the descending state; do not set `order='asc'` in the URL for ascending — represent ascending by the presence of `sort` only (no `order`). See the task description for expected URL param behavior .

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,275 @@
+  const sortField = params.get('sort');
+  const orderParam = params.get('order');
+  const sortOrder: 'asc' | 'desc' | null =
+    orderParam === 'desc' ? 'desc' : orderParam === 'asc' ? 'asc' : null;
```
When you read sort params you set sortOrder to `null` when the `order` param is absent. The task requires that ascending be represented by `sort=<field>` without `order`. That means if `sort` is present and `order` is absent, you must treat the state as ascending. Change the logic here to interpret missing `order` as `'asc'` when `sort` exists, e.g.:
`const sortOrder = sortField ? (orderParam === 'desc' ? 'desc' : 'asc') : null;`
If you don't do this, a URL like `#/people?sort=name` will be treated as unsorted or as descending by the current code, which violates the URL contract in the task description .

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,275 @@
-              </a>
-            </span>
+            <span className="is-flex is-align-items-center nowrap">Sex</span>
           </th>
```
The `Sex` column is currently a plain `<th>` (no SortHeader). The requirements explicitly require sorting by `sex` (one of: name, sex, born, died). Make the `Sex` header use `SortHeader` just like `Name`, `Born`, and `Died` so users can sort by sex and the `sort` param will be set accordingly.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,275 @@
+
+    const includes = (s?: string | null) =>
+      (s ?? '').toString().toLowerCase().includes(query.toLowerCase());
+
+    const filtered = people.filter(p => {
+      const matchesQuery =
+        !query ||
+        includes(p.name) ||
+        includes(p.motherName) ||
+        includes(p.fatherName);
+
+      const sexValue = String(p.sex ?? '')
+        .trim()
+        .toLowerCase();
+      const matchesSex =
+        !sexFilter ||
+        (sexFilter === 'm' && sexValue.startsWith('m')) ||
+        (sexFilter === 'f' && sexValue.startsWith('f'));
+
+      const bornCentury = Math.floor((p.born ?? 0) / 100) + 1;
+      const matchesCentury =
+        centuryFilters.length === 0 ||
+        centuryFilters.includes(String(bornCentury));
+
+      return matchesQuery && matchesSex && matchesCentury;
+    });
```
This component recalculates filtering (`people.filter(...)`) even though `PeoplePage` already computes `filteredPeople` and passes it in (see PeoplePage). Duplicated filtering increases maintenance burden and risks inconsistencies. Either accept the raw `people` list from parent and do filtering+sorting here, or accept the already-filtered list and only apply sorting. Decide on a single place for filtering and remove one of the duplicates to avoid divergence. See PeoplePage which currently computes `filteredPeople` from URL params .

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,275 @@
+                      : `#/people/${person.slug}`;
+
+                    // atualiza URL sem acionar o router que renderiza a outra página
+                    history.replaceState(null, '', newHash);
```
Row click navigation preserves current search params (you read them and append to the person slug, then `history.replaceState` and dispatch `hashchange`) — this matches the requirement to preserve search params when navigating within the People page. Good: keep this behavior. (No change required.)

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I’m requesting changes because a few core requirements aren’t met, which will break expected behavior.

Critical issues to fix:
- Navbar (`src/components/Navbar.tsx`): The People `NavLink` drops existing search params. Requirement: preserve query/sort params when navigating within People. Fix: import `useLocation` and set People link `to={`/people${location.search}`}` (e.g., around lines where `to="/people"` is defined).
- Sorting URL semantics (`src/components/PeopleTable.tsx`): The 3‑state sort cycle must use only `sort` and `order=desc`. Remove any `sortCycle` param and do not set `order=asc`. Implement cycle: 1) if switching field, set `sort=<field>` and remove `order`; 2) if same field and no `order`, set `order=desc`; 3) if same field and `order='desc'`, remove both `sort` and `order`.
- Interpret missing `order` as ascending when `sort` exists. Update the read logic so `sortOrder` becomes `'asc'` when `sort` is present and `order` is absent.
- Sex column not sortable: replace the plain `Sex` `<th>` with `SortHeader` (field `'sex'`) to meet the requirement to sort by name, sex, born, died.

Important improvements (non‑blocking but recommended):
- Avoid duplicated filtering between `PeoplePage` and `PeopleTable`; choose a single source of truth for filtering. Also consider centralizing hash search parsing (use a shared `readSearchParamsFromHash`).

You’ve done solid work on filters and preserving params in `PersonDetails`; addressing the above will bring the implementation fully in line with the requirements. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

DemetriodosAnjos committed:
feat(people): preservar query na navbar; 3-state sort sem sortCycle; coluna sex ordenável; util de hash

Criado util src/utils/hash.ts; Navbar preserva query ao navegar para /people; PeopleTable usa readSearchParamsFromHash/buildPeopleHashFromParams, interpreta sort sem order como asc, e implementa applySortExplicit(field) com ciclo de 3 estados sem sortCycle; SortHeader ajustado para nova assinatura e classes visuais. Testar lint/tsc e suite E2E.

DemetriodosAnjos committed:
fix(people): sincronizar seleção por slug; normalizar comparação; proteger timing

Adiciona util de hash; mantém applySortExplicit(field); força sincronização inicial do selectedSlug no mount; atualiza selectedSlug quando people chega; normaliza comparações de slug para evitar flakiness em CI; preserva query na Navbar.

DemetriodosAnjos committed:
fix(people): preserve slug in hash builder; sync selectedSlug on mount and when people arrive; add diagnostics logs

Preserve existing /people/<slug> when updating query params; ensure selectedSlug is set on mount and after people load to avoid CI timing issues; add console logs to help debugging flaky selection test.

DemetriodosAnjos committed:
revert(filters): restore PeopleFilters from 54aba32 (fim-etapa-ajustes-cores)

DemetriodosAnjos committed:
revert(filters): restore PeopleFilters from 8247ebc (fim-etapa-ajustes-cores)

DemetriodosAnjos committed:
fix(people): preserve query in navbar; implement 3-state sort usin only sort+order=desc  and treat missing order asc; make Sexo sortable; remove sortCycle and centralize hash handling.

DemetriodosAnjos committed:
fix(people): avoid rendering table when API returns people; restore filteredPeople in PeoplePage and preserve  no-results/no-people messaging

DemetriodosAnjos committed:
fix(people): do not render people table when API returns empty or on load error; let PeoplePage show no-results/no-people messages

mateacademy-ai-mentor commented on cypress/integration/page.spec.js:
```diff
@@ -7,47 +7,54 @@ const page = {
-    .should('have.length', 1)
-    .and('have.text', text),
-}
+  assetTitle: text =>
```
The helper in the `page` object is named `assetTitle` but it appears to be intended as an assertion helper (consistent with `assertHash` / `assertSearch`). Consider renaming to `assertTitle` (and update all usages) so intent is clear and consistent with other helpers.

mateacademy-ai-mentor commented on cypress/support/commands.ts:
```diff
@@ -47,17 +47,19 @@ declare global {
-Cypress.Commands.add('getByDataCy', selector => {
-  cy.get(`[data-cy="${selector}"]`);
+// cypress/support/commands.ts
+Cypress.Commands.add('getByDataCy', (selector: string) => {
+  return cy.get(`[data-cy="${selector}"]`);
```
The `getByDataCy` custom command returns `cy.get([data-cy="..."])` which is correct and matches the TS augmentation. Consider keeping it as-is; if you ever need access to Cypress `this` context prefer using a `function` instead of an arrow, but that's not necessary here. (See file where command is added)

mateacademy-ai-mentor commented on cypress/support/commands.ts:
```diff
@@ -47,17 +47,19 @@ declare global {
   'byDataCy',
   { prevSubject: 'optional' },
-
-  (subject, name) => {
+  (subject: any, name: string) => {
     const selector = `[data-cy="${name}"]`;
-
-    return subject ? cy.wrap(subject).find(selector) : cy.get(selector);
+    if (subject) {
+      return cy.wrap(subject).find(selector);
+    }
+    return cy.get(selector);
```
`byDataCy` implements a dual command (prevSubject: 'optional') and the runtime behavior is correct (wrap subject and find, otherwise global get). Two improvements:
1. Avoid `any` for `subject` — use `JQuery<HTMLElement> | undefined` (or a specific type) to retain TS safety.
2. The TypeScript augmentation above currently declares `byDataCy(name: string)` only as a single signature. To fully reflect the `prevSubject: 'optional'` usage you should add overloads in the `Cypress.Chainable` declaration so TypeScript knows the command can be used both as `cy.byDataCy('name')` and `cy.wrap(el).byDataCy('name')`.

This will prevent type errors when using the chainable both ways. (See command implementation)

mateacademy-ai-mentor commented on cypress/support/commands.ts:
```diff
@@ -47,17 +47,19 @@ declare global {
   'byDataCy',
   { prevSubject: 'optional' },
-
-  (subject, name) => {
+  (subject: any, name: string) => {
     const selector = `[data-cy="${name}"]`;
-
-    return subject ? cy.wrap(subject).find(selector) : cy.get(selector);
+    if (subject) {
+      return cy.wrap(subject).find(selector);
+    }
+    return cy.get(selector);
   },
 );
```
Small style / consistency note: you now have two helper names `getByDataCy` and `byDataCy`. Keep both if they are used in the test-suite, but if only one is used prefer a single canonical name to avoid confusion. Also ensure the declaration in the `declare global` block matches all overloads/usages.

mateacademy-ai-mentor commented on src/components/PeopleFilters.tsx:
```diff
@@ -1,16 +1,119 @@
+    readSearchParamsFromHash().getAll('centuries'),
+  );
+
+  useEffect(() => {
+    const onHashChange = () => {
+      const params = readSearchParamsFromHash();
+
+      setQuery(params.get('query') ?? '');
+      setSex(params.get('sex'));
+      setCenturies(params.getAll('centuries'));
+    };
+
+    window.addEventListener('hashchange', onHashChange);
+    onHashChange();
```
Good: initial state is read from the hash (so filters reflect the URL on load) and the component listens to hashchange to keep state in sync. This satisfies the requirement that filters reflect search params on navigation/refresh. (See the hash-change listener and initial reads)

mateacademy-ai-mentor commented on src/components/PeopleFilters.tsx:
```diff
@@ -1,16 +1,119 @@
+  const [sex, setSex] = useState<string | null>(() =>
+    readSearchParamsFromHash().get('sex'),
+  );
+  const [centuries, setCenturies] = useState<string[]>(() =>
+    readSearchParamsFromHash().getAll('centuries'),
+  );
+
+  useEffect(() => {
+    const onHashChange = () => {
+      const params = readSearchParamsFromHash();
+
+      setQuery(params.get('query') ?? '');
+      setSex(params.get('sex'));
+      setCenturies(params.getAll('centuries'));
+    };
+
+    window.addEventListener('hashchange', onHashChange);
+    onHashChange();
+
+    return () => window.removeEventListener('hashchange', onHashChange);
+  }, []);
+
+  useEffect(() => {
+    const params = readSearchParamsFromHash();
+
+    if (query.trim() === '') {
+      params.delete('query');
+    } else {
+      params.set('query', query.trim());
+    }
+
+    if (sex == null) {
+      params.delete('sex');
+    } else {
+      params.set('sex', sex);
+    }
+
+    params.delete('centuries');
+    centuries.forEach(c => params.append('centuries', c));
```
CenturyFilter multi-value semantics: the code uses `getAll('centuries')` to initialize and `params.append('centuries', c)` when writing the array back — this correctly implements the required `append` / `getAll` multi-value behaviour for centuries in the URL. Good. (See initialization and writing logic)

mateacademy-ai-mentor commented on src/components/PeopleFilters.tsx:
```diff
@@ -1,16 +1,119 @@
+    return () => window.removeEventListener('hashchange', onHashChange);
+  }, []);
+
+  useEffect(() => {
+    const params = readSearchParamsFromHash();
+
+    if (query.trim() === '') {
+      params.delete('query');
+    } else {
+      params.set('query', query.trim());
+    }
+
+    if (sex == null) {
+      params.delete('sex');
+    } else {
+      params.set('sex', sex);
+    }
+
+    params.delete('centuries');
+    centuries.forEach(c => params.append('centuries', c));
+
+    // buildPeopleHashFromParams garante remoção de sortCycle e preserva slug quando aplicável
+    const newHash = buildPeopleHashFromParams(params);
+
+    if (newHash !== window.location.hash) {
+      history.replaceState(null, '', newHash);
+      window.dispatchEvent(new HashChangeEvent('hashchange'));
+    }
+  }, [query, sex, centuries]);
```
Performance / event churn (optional): every keystroke triggers a history.replaceState + hashchange dispatch. This satisfies the requirement, but if you want to reduce the number of history updates you could debounce the query -> URL update; tests/requirements don't require this, so it's optional. (See the effect writing URL on `query` change)

mateacademy-ai-mentor commented on src/components/PeopleFilters.tsx:
```diff
@@ -1,16 +1,119 @@
+    });
+  };
+
+  const resetAll = () => {
+    setQuery('');
+    setSex(null);
+    setCenturies([]);
+  };
```
Reset behaviour: `resetAll` sets state to defaults and the existing effect will remove `query`, `sex`, and `centuries` from the URL. This correctly implements the Reset all filters requirement. (See resetAll and the Reset anchor)

mateacademy-ai-mentor commented on src/components/PeopleFilters.tsx:
```diff
@@ -33,52 +137,46 @@ export const PeopleFilters = () => {
-            <a
-              data-cy="century"
-              className="button mr-1"
-              href="#/people?centuries=20"
-            >
-              20
-            </a>
+                  p.delete('centuries');
+                  if (willSelect) {
+                    p.append('centuries', c);
+                    current
+                      .filter(x => x !== c)
+                      .forEach(x => p.append('centuries', x));
+                  } else {
+                    current
+                      .filter(x => x !== c)
+                      .forEach(x => p.append('centuries', x));
+                  }
```
Implementation clarity suggestion: the century `href` builder uses a logic that reconstructs the `centuries` param ordering (it deletes then appends). That is fine — ensure other parts of the app expect repeated `centuries` params in any order. There is no functional bug here, only a note to be consistent across the app. (See century href builder)

mateacademy-ai-mentor commented on src/components/PeopleFilters.tsx:
```diff
@@ -1,16 +1,119 @@
+
+  const centurySet = useMemo(() => new Set(centuries), [centuries]);
+
+  const hrefWith = (changes: (p: URLSearchParams) => void) => {
+    const p = readSearchParamsFromHash();
+
+    changes(p);
+
+    return buildPeopleHashFromParams(p);
+  };
```
Small type-safety suggestion (optional): `readSearchParamsFromHash()` and `buildPeopleHashFromParams()` are used liberally and work well. If you want stronger TypeScript checks, consider typing the `changes` callback used in `hrefWith` more explicitly (e.g., `(p: URLSearchParams) => void`) — it's already that shape but an explicit interface helps readability. This is optional and not required for tests. (See `hrefWith`)

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
+  const [query, setQuery] = useState<string>('');
+  const [centuryFilters, setCenturyFilters] = useState<string[]>([]);
+
+  useEffect(() => {
+    const applyParams = () => {
+      const params = readSearchParamsFromHash();
+
+      setQuery(params.get('query') ?? '');
+      setSexFilter(params.get('sex'));
+      setCenturyFilters(params.getAll('centuries'));
```
Reading search params on mount and on `hashchange` and updating local filters is implemented correctly. This matches the requirement to keep the page state in sync with the URL so the view can be shared and navigation preserves filters/sort. Good use of `readSearchParamsFromHash` and state setters.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
+    return () => window.removeEventListener('hashchange', applyParams);
+  }, []);
+
+  const filteredPeople = useMemo(() => {
+    if (!people) {
+      return [];
+    }
+
+    const includes = (s?: string | null) =>
+      (s ?? '').toString().toLowerCase().includes(query.toLowerCase());
+
+    return people.filter(p => {
+      const matchesQuery =
+        !query ||
+        includes(p.name) ||
+        includes(p.motherName) ||
+        includes(p.fatherName);
+
+      const sexValue = String(p.sex ?? '')
+        .trim()
+        .toLowerCase();
+      const matchesSex =
+        !sexFilter ||
+        (sexFilter === 'm' && sexValue.startsWith('m')) ||
+        (sexFilter === 'f' && sexValue.startsWith('f'));
+
+      const bornCentury = Math.floor((p.born ?? 0) / 100) + 1;
+      const matchesCentury =
+        centuryFilters.length === 0 ||
+        centuryFilters.includes(String(bornCentury));
+
+      return matchesQuery && matchesSex && matchesCentury;
+    });
+  }, [people, query, sexFilter, centuryFilters]);
```
Critical: duplicated filtering. This file computes `filteredPeople` (lines 34–64) and then passes `filteredPeople` as the `people` prop into `PeopleTable` (lines 104–110). But `PeopleTable` also performs filtering based on `query`, `sexFilter` and `centuryFilters` (see `src/components/PeopleTable.tsx`). Having the same filter applied in two places leads to redundancy and potential bugs / unexpected behavior. Pick one place to perform filtering and keep the other one out of the loop.

Recommended options:
- Move all filtering into `PeopleTable`: pass the raw `people` (the `people` prop) from PeoplePage and remove `filteredPeople` here; keep `query`, `sexFilter`, `centuryFilters` as props or let PeopleTable read params itself.
- Or keep filtering here and remove filtering logic from `PeopleTable` so it only renders the provided list.

Either way, keep a single source of truth for filtering to avoid subtle mismatches. See PeopleTable where it also filters/sorts.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
+              !loading &&
+              people &&
+              filteredPeople.length > 0 && (
+                <PeopleTable
+                  people={filteredPeople}
+                  loading={loading}
+                  error={error}
+                  sexFilter={sexFilter}
+                  query={query}
+                  centuryFilters={centuryFilters}
```
Consequence of the duplicated filtering: currently you pass `filteredPeople` into the table at lines 104–110. If you choose to centralize filtering in `PeopleTable`, change these props to pass the raw `people` and (optionally) the `sexFilter`, `query`, `centuryFilters` values, or remove those props if `PeopleTable` will read params itself. If you choose to keep filtering here, ensure `PeopleTable` does not reapply filters. This keeps behavior predictable and makes it easier to reason about URL→state synchronization.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -1,32 +1,124 @@
           </div>
         </div>
+
+        {people && (
+          <div className="column is-12-mobile is-12-tablet is-4-desktop">
+            <PeopleFilters />
```
Small suggestion about showing the filters sidebar. You render `PeopleFilters` when `people` is truthy (lines 117–119). This means the sidebar appears after `people` changes from `null` to an array (which indicates data loaded) — that satisfies the requirement to show filters only after data is loaded. If you want to be explicit and avoid showing filters while data is still loading, consider `!loading && people !== null` as the condition. Decide based on whether you want the sidebar visible when the API returns an empty array (currently it will be visible).

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+
+      return matchesQuery && matchesSex && matchesCentury;
+    });
+
+    if (!sortField) {
+      return filtered;
+    }
+
+    return [...filtered].sort((a, b) => {
+      const getValue = (person: Person) => {
+        switch (sortField) {
+          case 'name':
+            return person.name ?? '';
+          case 'sexo':
+            return person.sex ?? '';
+          case 'born':
+            return person.born ?? 0;
+          case 'died':
+            return person.died ?? 0;
+          case 'mother':
+            return person.motherName ?? '';
+          case 'father':
+            return person.fatherName ?? '';
+          default:
+            return '';
+        }
+      };
+
+      const aVal = getValue(a);
+      const bVal = getValue(b);
+
+      if (typeof aVal === 'number' && typeof bVal === 'number') {
+        return sortOrder === 'desc' ? bVal - aVal : aVal - bVal;
+      }
+
+      return sortOrder === 'desc'
+        ? String(bVal).localeCompare(String(aVal))
+        : String(aVal).localeCompare(String(bVal));
+    });
+  }, [people, query, sexFilter, centuryFilters, sortField, sortOrder]);
```
Duplicate filtering: this file computes `filteredPeople` (the full filtering pipeline) in the component (lines 119–185) but PeoplePage also computes `filteredPeople` and passes it into PeopleTable. Keeping the same filtering logic in two places will cause confusion and may double-filter the list or hide intended results. Choose one place to perform filtering. Recommended approaches:
- Move all filtering into PeopleTable: accept the raw `people` prop (unfiltered) from PeoplePage and remove `filteredPeople` from PeoplePage, or
- Keep filtering in PeoplePage and remove filtering from PeopleTable (PeopleTable becomes a pure presentational table that only sorts the passed list).

The project tests and requirements expect search params to be persisted and applied consistently — centralizing filtering will help. See how PeoplePage reads URL params and computes its own `filteredPeople` (PeoplePage usage: it currently passes `filteredPeople` to PeopleTable).

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+        switch (sortField) {
+          case 'name':
+            return person.name ?? '';
+          case 'sexo':
```
Incorrect sort field name for sex: the sorting `getValue` uses `case 'sexo': return person.sex` (line 159) and the `SortHeader` is rendered with `field="sexo"` (line 211). The requirements and the Person type use `sex`, and the `sort` URL param should be `sort=sex`. Use `sex` consistently for both `SortHeader` field and the switch case so the URL and sorting behavior match the requirement. Update both places from `'sexo'` to `'sex'`.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+            applySortExplicit={applySortExplicit}
+          />
+          <SortHeader
+            field="sexo"
```
Also update the SortHeader usage to `field="sex"` (currently `field="sexo"` at the header invocations). Without this change clicking the sex column will set `?sort=sexo`, which does not match the expected `sort=sex` semantics. Update the header field argument at line 211.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+  return params.get('selected');
+}
+
+function applySortExplicit(field: string) {
+  const params = readSearchParamsFromHash();
+  const currentField = params.get('sort');
+  const currentOrder = params.get('order');
+
+  if (currentField !== field) {
+    params.set('sort', field);
+    params.delete('order');
+  } else {
+    if (!currentOrder) {
+      params.set('order', 'desc');
+    } else if (currentOrder === 'desc') {
+      params.delete('sort');
+      params.delete('order');
+    } else {
+      params.set('order', 'desc');
+    }
+  }
+
+  params.delete('sortCycle');
+
+  const newHash = buildPeopleHashFromParams(params);
+
+  if (newHash !== (typeof window !== 'undefined' ? window.location.hash : '')) {
+    history.replaceState(null, '', newHash);
+    window.dispatchEvent(new HashChangeEvent('hashchange'));
+  }
+}
```
Sorting URL semantics / cycle: `applySortExplicit` (lines 34–61) mostly implements the required 3-state cycle and writes only `order=desc` when switching to descending — this is correct. Keep the cycle strictly as: (none -> set `sort=<field>` and delete `order`) -> (same field and no `order` -> set `order=desc`) -> (same field and `order=desc` -> delete both `sort` and `order`).

One small note: you call `params.delete('sortCycle')` at line 53 — make sure no other code writes `sortCycle`. Also the read logic later treats `order==='asc'` (lines 104–114) — per requirements `order=asc` should not be written; you can keep the read branch for robustness but ensure no writer sets `order=asc`. The main requirement is to write only `sort` (for ascending) and `order=desc` (for descending) and to remove both when disabling sorting. The implementation in `applySortExplicit` adheres to that, but verify no other components write `order=asc`. Requirements source: sorting must use only `sort` and `order=desc`.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [people]);
+
+  const params = readSearchParamsFromHash();
+  const sortField = params.get('sort');
+  const orderParam = params.get('order');
+  const sortOrder: 'asc' | 'desc' | null =
+    sortField && !orderParam
+      ? 'asc'
+      : orderParam === 'desc'
+        ? 'desc'
+        : orderParam === 'asc'
+          ? 'asc'
+          : null;
```
Reading sort/order: the code interprets a missing `order` as ascending when `sort` exists (lines 104–114) — this matches the requirement (good). However, since `order=asc` should never be written, the `order === 'asc'` branch (line 112–113) is unnecessary; it is harmless but may be removed to simplify logic. Keep interpretation of missing `order` as `'asc'`.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+  }, [people, query, sexFilter, centuryFilters, sortField, sortOrder]);
+
+  // não renderizar tabela quando não há dados carregados, quando a lista está vazia ou quando há erro
+  if (
+    !people ||
+    (Array.isArray(people) && people.length === 0) ||
+    Boolean(error)
+  ) {
+    return null;
```
Early return hides table when `people` is falsy / empty / error (lines 188–193). Note: when you centralize filtering (see first comment) ensure that the table's render conditions are correct for the chosen flow. For example, if PeoplePage stops passing `filteredPeople` and passes raw `people`, PeopleTable should still render (and show no-results behavior) when the filtered list is empty. Decide where you want to show the 'no results' message (PeoplePage currently shows it) and keep that responsibility in one place to avoid UI inconsistencies.

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
+  buildPeopleHashFromParams,
+} from '../utils/hash';
+
+type Props = {
+  people: Person[] | null;
+  loading: boolean;
+  error: string | null;
+  sexFilter: string | null;
+  query: string;
+  centuryFilters: string[];
+};
```
Props: `loading` is declared in Props (line 13) but not used in this component. If you don't need `loading` inside PeopleTable, remove it from the props type to avoid unused prop noise. If you will need it (for example to show a loader inside table) either use it or annotate why it is present. Keeping the props type minimal helps maintainability. (Props declaration at lines 11–18).

mateacademy-ai-mentor commented on src/components/PeopleTable.tsx:
```diff
@@ -1,645 +1,300 @@
-          </td>
-        </tr>
+              <td>
+                <a
+                  className={`is-clickable ${nameClass}`}
+                  onClick={() => {
+                    const clickParams = readSearchParamsFromHash();
+
+                    clickParams.delete('page');
+                    const qs = clickParams.toString();
+                    const newHash = qs
+                      ? `#/people/${person.slug}?${qs}`
+                      : `#/people/${person.slug}`;
+
+                    history.replaceState(null, '', newHash);
+                    window.dispatchEvent(new HashChangeEvent('hashchange'));
+                    setSelectedSlug(person.slug);
```
Person selection / preserving search params when selecting a person: the click handler on the person name (lines 268–281) uses `readSearchParamsFromHash()` and preserves all current params when building the new hash for `/people/<slug>?<qs>` — that satisfies the requirement to keep search params while navigating within People. Good pattern to preserve filters/sort when selecting a person. Keep this behavior when you move filtering/sorting responsibilities.

mateacademy-ai-mentor commented on src/components/PersonDetails.tsx:
```diff
@@ -0,0 +1,116 @@
+  const { slug } = useParams<{ slug: string }>();
+  const [searchParams] = useSearchParams();
+
+  if (people === null) {
+    return <p data-cy="personLoading">Loading person...</p>;
```
Good: the component handles the loading state (`people === null`) and not-found state properly and uses `useSearchParams` to read current params. This matches the requirement to preserve filters/sort when navigating and to avoid rendering person details before data is loaded. (See lines that return `personLoading` and `personNotFound`.)

mateacademy-ai-mentor commented on src/components/PersonDetails.tsx:
```diff
@@ -0,0 +1,116 @@
+
+  const bornCentury = Math.floor(person.born / 100) + 1;
+
+  // Interpret missing order as 'asc' when a sort field is present
+  const sortParam = searchParams.get('sort') ?? null;
+  const rawOrder = searchParams.get('order');
+  const orderParam = sortParam && !rawOrder ? 'asc' : (rawOrder ?? null);
```
Critical: you set `orderParam` to `'asc'` when `sort` is present but no `order` param exists (line 30). That causes `preservedParams` to include `order: 'asc'` and will make SearchLink write `order=asc` into links. The task requires that the URL include `order=desc` only for descending sorts and omit `order` for ascending. Adjust the logic so you still *interpret* missing `order` as ascending for internal use, but do NOT include `order='asc'` in `preservedParams`. For example, keep `orderForRead = sortParam && !rawOrder ? 'asc' : rawOrder ?? null` for internal interpretation, but when building preserved params only include `order: 'desc'` (i.e., `rawOrder === 'desc' ? 'desc' : null`).

mateacademy-ai-mentor commented on src/components/PersonDetails.tsx:
```diff
@@ -0,0 +1,116 @@
+  const rawOrder = searchParams.get('order');
+  const orderParam = sortParam && !rawOrder ? 'asc' : (rawOrder ?? null);
+
+  const preservedParams = {
+    query: searchParams.get('query') ?? null,
+    sex: searchParams.get('sex') ?? null,
+    centuries: searchParams.getAll('centuries'),
+    sort: sortParam,
+    order: orderParam,
```
Related: `preservedParams.query` is built using `searchParams.get('query') ?? null` (line 33). If the `query` param exists but is an empty string this will keep `query: ''` in `preservedParams` and SearchLink will write an empty `query` param into the URL. The requirement states that when the NameFilter input is empty there should be no `query` param. Change this to map empty strings to `null` (e.g. `const rawQ = searchParams.get('query'); query: rawQ && rawQ.trim() ? rawQ : null`) so empty input results in the `query` key being removed. Also apply the same defensive rule if you expect any other params that could be empty.

mateacademy-ai-mentor commented on src/components/PersonDetails.tsx:
```diff
@@ -0,0 +1,116 @@
+  const rawOrder = searchParams.get('order');
+  const orderParam = sortParam && !rawOrder ? 'asc' : (rawOrder ?? null);
+
+  const preservedParams = {
+    query: searchParams.get('query') ?? null,
+    sex: searchParams.get('sex') ?? null,
+    centuries: searchParams.getAll('centuries'),
+    sort: sortParam,
+    order: orderParam,
+  };
```
Minor: `preservedParams` includes `sort: sortParam` which is fine (it will be removed by the helper when `null`). After you change the `order` handling to only include `desc` when appropriate and avoid empty `query`, the SearchLink usages (mother/father/back link) will preserve params exactly as required by the spec. Ensure other parts of the app that *write* params (filters, sort toggles) use the same conventions (only `order=desc` when descending; omit `order` for ascending).

mateacademy-ai-mentor commented on src/utils/hash.ts:
```diff
@@ -0,0 +1,51 @@
+ * Caso contrário retorna "#/people" ou "#/people?..." dependendo de params.
+ * Antes de serializar, remove parâmetros residuais que não devem ser persistidos, como sortCycle.
+ */
+export function buildPeopleHashFromParams(params: URLSearchParams): string {
+  // garantir remoção de parâmetros não desejados
+  params.delete('sortCycle');
```
Minor robustness: `buildPeopleHashFromParams` mutates the incoming `params` by calling `params.delete('sortCycle')` (line 22). Mutating a caller-supplied URLSearchParams can surprise callers. Consider cloning the params at the start (e.g. `const p = new URLSearchParams(params.toString())`) and operate on `p` so the original object is not modified by this function. This will make the function safer for reuse by other code (PeopleFilters/PeopleTable rely on these helpers).

mateacademy-ai-mentor commented on src/utils/hash.ts:
```diff
@@ -0,0 +1,51 @@
+
+    return qs ? `#/people?${qs}` : '#/people';
+  }
+
+  const currentHash = window.location.hash || '';
+  const base = currentHash.split('?')[0] || '';
+  const parts = base
+    .split('/')
+    .map(p => p.trim())
+    .filter(Boolean);
```
Parsing robustness: currently you take `currentHash = window.location.hash` and then `base = currentHash.split('?')[0]` and split that by '/'. That leaves the leading `#` as part of the first `parts` item (e.g. `'#'`), which is harmless in current logic but makes the `parts` array include `'#'`. To make the slug extraction clearer and avoid subtle bugs, strip the leading `#` before splitting (e.g. `const base = (currentHash.split('?')[0] || '').replace(/^#/, '')`). This makes `parts` contain only path segments and avoids a `'#'` element in `parts`. The relevant lines are 29–35.

mateacademy-ai-mentor commented on src/utils/hash.ts:
```diff
@@ -0,0 +1,51 @@
+  // garantir remoção de parâmetros não desejados
+  params.delete('sortCycle');
+
+  if (typeof window === 'undefined') {
+    const qs = params.toString();
+
+    return qs ? `#/people?${qs}` : '#/people';
```
Correct behavior note: when no `window` is available you return a string built from `params.toString()` (`#/people?${qs}` or `#/people`) which is correct for SSR or tests that call the function without a DOM. Good coverage for both environments (lines 24–27).

mateacademy-ai-mentor commented on src/utils/hash.ts:
```diff
@@ -0,0 +1,51 @@
+    return new URLSearchParams();
+  }
+
+  const hash = window.location.hash || '';
+  const idx = hash.indexOf('?');
+  const qs = idx >= 0 ? hash.slice(idx + 1) : '';
+
+  return new URLSearchParams(qs);
```
Behavior check: `readSearchParamsFromHash` extracts the query-string portion after `?` into a new URLSearchParams which is exactly what other modules expect (e.g., `getAll('centuries')`, `get('sort')`). This makes it straightforward for components like PeopleFilters and PeopleTable to use `append`/`getAll` semantics. No change needed here (lines 7–11).

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I’m requesting changes because there are a few critical blockers that violate required URL/filtering semantics and can break expected behavior.

Critical issues to fix:
- Duplicate filtering logic in two places: `src/components/PeoplePage.tsx` (lines 34–64) and `src/components/PeopleTable.tsx` (lines 119–185). This can double-filter and produce unexpected results. Choose a single source of truth (recommend centralize in one component) and remove the other.
- Wrong sort field key for Sex in `src/components/PeopleTable.tsx`: `getValue` uses `'sexo'` (around line 159) and headers pass `field="sexo"` (around line 211). The requirement is `sort=sex`. Rename both to `sex` so sorting and URLs work.
- Incorrect preserved URL params in `src/components/PersonDetails.tsx`: when `sort` exists and `order` is missing, you set `order='asc'` (around line 30). The spec requires omitting `order` for ascending and using `order=desc` only for descending. Also, ensure `query` is omitted when empty instead of preserving `query=''` (around line 33).

Minor/non-blocking improvements:
- Cypress spec helper typo: `assetTitle` → `assertTitle` in `cypress/integration/page.spec.js`.
- In `cypress/support/commands.ts`, add correct TypeScript overloads and avoid `any` for the dual `byDataCy` command; consider consolidating command names if only one is used.

You’ve done solid work on routing, URL param helpers, and preserving params through navigation. Addressing the above will align the app with the required URL semantics and make behavior predictable. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.
